### PR TITLE
Track Source Code of Steps

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
We now track the source code of each step (so we can show it in the dashboard). If the source code is longer than 1000 chars, we truncate it to reduce storage space. 

Similarly, the `docstring` field of steps now also gets truncated if it is longer than 1000 chars and it is no longer contained in `config.Step`.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by strickvl — COMMENTED** [review]
This looks good / clean to me. I think we've set the maximum text size that can be entered to the absolute maximum, so we should be good here.

Is there a post-execution-style test we could add in here to make sure that this new feature works as hoped, and perhaps also a test to denote the limit of how much code you can stuff into the database?

Also, I'm thinking this can / should go somewhere in the docs that this is a feature of what we're keeping track of, and what can be accessed on the dashboard.

File: src/zenml/steps/step_decorator.py [hunk 1]
```diff
@@ -146,6 +148,7 @@ def inner_decorator(func: F) -> Type[BaseStep]:
                     PARAM_OUTPUT_MATERIALIZERS: output_materializers,
                     PARAM_SETTINGS: settings,
                     PARAM_EXTRA_OPTIONS: extra,
+                    PARAM_SOURCE_CODE: inspect.getsource(func),
```
Comment (htahir1) [inline src/zenml/steps/step_decorator.py:151]: Do we want to encode this to compress it / make safer?

**Review by htahir1 — CHANGES_REQUESTED** [review]

File: src/zenml/steps/base_step.py [hunk 2]
```diff
@@ -260,6 +261,7 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
 
         kwargs = {**self.INSTANCE_CONFIGURATION, **kwargs}
         name = kwargs.pop(PARAM_STEP_NAME, None) or self.__class__.__name__
+        source_code = kwargs.pop(PARAM_SOURCE_CODE, None)
```
Comment (schustmi) [inline src/zenml/steps/base_step.py:264]: This does not work for class-based steps now right? You can take a look at `BaseStep.caching_parameters` to see how we handle it for both

**Review by schustmi — CHANGES_REQUESTED** [review]
Some general comments:
- I agree with @strickvl that we should have some way to catch overflowing fields. It should fail with a reasonable error message or even discard fields like the docstring/code to make it fit into the database limits.
- This also unnecessarily "bloats" the StepConfiguration which is part of the intermediate representation that we serialize. We could instead also get this information when writing the database entry, at which point I think we re-create the `BaseStep` instance anyway
- This makes the `docstring` field superfluous I think. The `inspect.getsource` call includes the docstring, are we going to visualize both?

**Comment by fa9r** [comment]
`docstring` field is redundant now, yes. But it is still useful I think, e.g., if you want to display the docstring specifically as a description in the UI or similar. Not trivial to get that from the full source code, probably you'd need to write some custom regex for that

**Comment by fa9r** [comment]
Regarding DB overflow, I wanted to add memory checks but it's not so trivial since the code is saved as part of the step config. I like your idea of dropping code/docstring though, will add some logic that does that in case the step config becomes too big

File: src/zenml/steps/step_decorator.py (hunk 1 above)
Comment (fa9r) [reply src/zenml/steps/step_decorator.py]: Good idea, added `zlib` encoding for the `zenml.config.Step` object in general.

File: src/zenml/steps/base_step.py (hunk 2 above)
Comment (fa9r) [reply src/zenml/steps/base_step.py]: Changed it to also work for class-based steps. This makes it much cleaner in general, great suggestion 👍

**Comment by fa9r** [comment]
@schustmi @strickvl @htahir1 when writing a step to DB, we now encode the step configuration to reduce storage space. If that is not either, we truncate the `docstring` and `source_code` fields.

I also added two post-execution tests that ensure that `source_code` is set correctly and that it gets truncated if its too long.

File: tests/integration/functional/post_execution/test_pipeline_run.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#  Copyright (c) ZenML GmbH 2021. All Rights Reserved.
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
```
Comment (htahir1) [inline tests/integration/functional/post_execution/test_pipeline_run.py:1]: 👀

**Review by htahir1 — CHANGES_REQUESTED** [review]
Great! About compressing before database insertion I looked around on second thought as  imagine the underlying databases are already compressing the texts.. And I found this thread: https://stackoverflow.com/questions/8228950/compressing-text-before-storing-it-in-the-database

Apparently there is no real need for compressing them before entry as it makes the database less searchable. However if one does compress, then dont use TEXT to store it, rather BLOG, as the bytes might get mangled

**Comment by fa9r** [comment]
Will wait for more reviews before removing the encoding again

File: src/zenml/config/step_configurations.py [hunk 4]
```diff
@@ -187,3 +194,48 @@ class Step(StrictBaseModel):
...
+
+    def encode(self) -> bytes:
+        """Encodes the step into a byte string.
+
+        Returns:
+            The encoded step.
+        """
+        step_dict = self.dict(exclude_none=True)
+        step_json = json.dumps(step_dict, sort_keys=True)
+        encoded = zlib.compress(step_json.encode())
```
Comment (strickvl) [inline src/zenml/config/step_configurations.py:206]: Wonder whether it's worth being explicit about the level of compression here? Not quite sure where we are on the scale of speed:compression, but I don't think it's bang in the middle... https://docs.python.org/3/library/zlib.html#zlib.compress

**Review by strickvl — APPROVED** [review]
This looks good to me now. I had one question around the extent to which we compress, but feel free to ignore that. Perhaps defaults just are best.

**Review by schustmi — COMMENTED** [review]
Code-wise this looks fine to me, however I still have my original concern of this bloating the intermediate representation. The `Step` class was never intended to be a catch-all class to add random stuff to, but to contain all the necessary information needed to run a step. Now that it gets added to the database, we abuse it by adding stuff like the code and docstring.
I'd also suggest doing some tests how much benefit the encoding/compression really gives us, and whether readability/searchability isn't actually more useful?

File: src/zenml/steps/step_decorator.py (hunk 1 above)
Comment (fa9r) [reply src/zenml/steps/step_decorator.py]: Removed encoding/decoding again following reviewer consensus

File: src/zenml/config/step_configurations.py (hunk 4 above)
Comment (fa9r) [reply src/zenml/config/step_configurations.py]: Removed encoding again according to reviewer consensus

**Comment by fa9r** [comment]
@schustmi I removed both `docstring` and `source_code` from the step config as discussed.

File: src/zenml/orchestrators/step_launcher.py [hunk 5]
```diff
@@ -186,6 +189,28 @@ def launch(self) -> None:
...
+
+        Returns:
+            The docstring and source code of the step.
+        """
+        from zenml.steps.base_step import BaseStep
+
+        step_instance = BaseStep.load_from_source(self._step.spec.source)
+
+        docstring = step_instance.docstring
+        if docstring and len(docstring) > 1000:
```
Comment (schustmi) [inline src/zenml/orchestrators/step_launcher.py:205]: Shouldn't this be `TEXT_FIELD_MAX_LENGTH<3` or something? 1000 seems quite a low amount of characters for code/docstring?

File: src/zenml/steps/base_step.py [hunk 6]
```diff
@@ -322,6 +323,21 @@ def _created_by_functional_api(cls) -> bool:
             PARAM_CREATED_BY_FUNCTIONAL_API, False
         )
 
+    @classmethod
```
Comment (schustmi) [inline src/zenml/steps/base_step.py:326]: Love this!

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/orchestrators/step_launcher.py (hunk 5 above)
Comment (fa9r) [reply src/zenml/orchestrators/step_launcher.py:205]: I thought about that but intentionally kept it at 1000. The use case of this is only to have something to display in the UI, I don't think it's worth potentially saving 65KB of data per step for that

**Review by schustmi — APPROVED** [review]

**Pull request merged**